File: day9.c
#include <stdio.h>
#include <string.h>

char buf[20000 + 1];
int map[200000];
int n;

void printmap()
{
    for (int i = 0; i < n; i++) {
        if (map[i] >= 0) {
            printf("%d", map[i]);
        } else {
            printf(".");
        }
    }
    printf("\n");
}

int main()
{
    FILE* file = fopen("day9.txt", "r");
    long long sum = 0, sum2 = 0;

    fscanf(file, "%s", buf);
    int buflen = strlen(buf);

    int nextid = 0;
    for (int i = 0; i < buflen; i++) {
        int l = buf[i] - '0';
        int id;
        if (i & 1) id = -1;
        else id = nextid++;
        for (int j = 0; j < l; j++) {
            map[n++] = id;
        }
    }

    //printmap();

    for (int i = 0, j = n - 1; i <= j; i++) {
        if (map[i] >= 0) {
            sum += i * map[i];
        } else {
            while (j > i && map[j] < 0) {
                j--;
            }
            if (j <= i) break;
            sum += i * map[j];
            j--;
        }
    }

    for (int sbeg = n - 1; sbeg >= 0; ) {
        while (sbeg >= 0 && map[sbeg] < 0) {
            sbeg--;
        }
        if (sbeg < 0) break;
        int id = map[sbeg];
        int send = sbeg - 1;
        while (send >= 0 && map[send] == id) {
            send--;
        }
        int slen = sbeg - send;
        int dbeg = 0, dlen = 0;
        while (dbeg <= send) {
            while (dbeg <= send && map[dbeg] >= 0) {
                dbeg++;
            }
            if (dbeg > send) break;
            int dend = dbeg + 1;
            while (dend < n && map[dend] < 0) {
                dend++;
            }
            dlen = dend - dbeg;
            if (dlen >= slen) break;
            dbeg = dend;
            dlen = 0;
        }
        if (dlen >= slen) {
            for (int i = 0; i < slen; i++) {
                map[dbeg + i] = id;
                map[send + 1 + i] = -1;
            }
        }
        sbeg = send;
    }

    for (int i = 0; i < n; i++) {
        if (map[i] >= 0) {
            sum2 += i * map[i];
        }
    }

    //printmap();

    printf("%lld %lld", sum, sum2);
}

File: day9.lua
#!/usr/bin/env lua

local file = io.open("day9.txt")
local sum, sum2 = 0, 0

local buf = file:read("l")
local map = {}
local files = {}
local spaces = {}

local pos = 0
for i = 1, #buf do
    local l = tonumber(buf:sub(i, i))
    local id = i & 1 ~= 0 and i // 2 or -1
    for j = 1, l do
        table.insert(map, id)
    end
    table.insert(i & 1 ~= 0 and files or spaces, { pos, l })
    pos = pos + l
end

local i, j = 1, #map
while i <= j do
    if map[i] >= 0 then
        sum = sum + (i - 1) * map[i]
    else
        while j > i and map[j] < 0 do
            j = j - 1
        end
        if j <= i then break end
        sum = sum + (i - 1) * map[j]
        j = j - 1
    end
    i = i + 1
end

for i = #files, 1, -1 do
    local fpos, flen = table.unpack(files[i])
    for j = 1, #spaces do
        local spos, slen = table.unpack(spaces[j])
        if spos >= fpos then break end
        if flen <= slen then
            files[i] = { spos, flen }
            spaces[j] = { spos + flen, slen - flen }
            break
        end
    end
end

for i, file in ipairs(files) do
    local pos, len = table.unpack(file)
    for j = 0, len - 1 do
        sum2 = sum2 + (i - 1) * (pos + j)
    end
end

print(sum, sum2)
